### DISCUSSION TITLE
Smagorinsky closure with a damping function

### DISCUSSION THREAD

**Discussion by j-ostiguy** [CliMA/Oceananigans.jl#3369] [Computational science!]
Hello,

I would like to use the Smagorinsky closure, but where the length scale includes a van Driest damping function (i.e. l = C*Δ*[1-exp(-y+/A+)] ). Is there a way to do this? 

Thank you!

**Comment by glwagner** [comment]
The length scale (which we call the "filter width") is implemented here:

https://github.com/CliMA/Oceananigans.jl/blob/f3996c2703bb23b2e84f80c6f1564fc7965bdad1/src/TurbulenceClosures/turbulence_closure_implementations/smagorinsky_lilly.jl#L114-L116

To generalize this requires a bit of source code development (not much...)

Another route is to generalize the "model constant", which multiplies the length-scale, eg

https://github.com/CliMA/Oceananigans.jl/blob/f3996c2703bb23b2e84f80c6f1564fc7965bdad1/src/TurbulenceClosures/turbulence_closure_implementations/smagorinsky_lilly.jl#L117-L119

We've provided such a feature for AMD:

https://github.com/CliMA/Oceananigans.jl/blob/f3996c2703bb23b2e84f80c6f1564fc7965bdad1/src/TurbulenceClosures/turbulence_closure_implementations/anisotropic_minimum_dissipation.jl#L140-L144

What are `Δ`, `C`, `y+`, and `A+` in your expression? Personally, I think it makes sense to support this kind of thing by generalizing the model constant calculation as we've done for AMD.

**Reply by j-ostiguy** [reply]
Δ is the filter width, C is the Smagorinksy constant (~0.16), y+=y/δ_ν is the distance from the wall/boundary in viscous lengths, and A+ is a constant (usually given the value 26). 

I think generalizing the model constant would work great since the idea is to set l as above instead of l=C*Δ.

**Reply by glwagner** [reply]
I'm thinking that we should stop supporting `Function` here and instead encourage people to utilize `FunctionField`, which is pretty flexible anyways.

What is `δ_ν`? Viscosity has units of length and time, so we'd need a time unit to convert viscosity into a length scale.

(I'm asking because I want to make sure our user interface supports common usage in a sensible way that facilitates good communication of concepts.)

Distance-to-wall has to be defined by the user I think. It's a little annoying for irregular boundaries. We can get to that.

> A+ is a constant (usually given the value 26).

👀 heh, nice!

**Reply by glwagner** [reply]
I'll do something simple to start to illustrate how these kinds of things are done. But people who use this feature should feel free to implement a more general and sophisticated interface if they want that.

**Reply by glwagner** [reply]
Let's move the discussion to https://github.com/CliMA/Oceananigans.jl/pull/3370

**Reply by j-ostiguy** [reply]
> What is δ_ν? Viscosity has units of length and time, so we'd need a time unit to convert viscosity into a length scale.

δ_ν = ν*sqrt(ρ/ τ_0) = ν/u_τ  is the viscous length scale (where τ_0 is the stress at the wall and u_τ is the friction velocity). I'm not sure I understand why viscosity has to be converted to a length scale. 

>Distance-to-wall has to be defined by the user I think. It's a little annoying for irregular boundaries. We can get to that.

y+, the distance to the wall, can be written as y+ = u_τ*y/ν. So, it's a function of the shear stress at the boundary - can you get that from the model? Or I could calculate u_τ from the momentum flux at the boundary.

**Reply by glwagner** [reply]
> δ_ν = ν*sqrt(ρ/ τ_0) = ν/u_τ is the viscous length scale (where τ_0 is the stress at the wall and u_τ is the friction velocity). I'm not sure I understand why viscosity has to be converted to a length scale.

Maybe I miscommunicated, but `u_τ` is precisely the scale (here a velocity scale) that mediates between viscosity and the length scale `δ_ν`. So you provided what I asked.

If `u_τ` is not prescribed by forcing, then we would have to estimate the friction velocity dynamically based on the near-wall velocity gradient, right? If we need to do that, things are a little tricker (we probably want to compute the friction velocity within a callback). Let me know if you'd like me to show you how to do that. I provided an example in the PR that assumes the near-wall length scale is known a priori.

**Reply by glwagner** [reply]
> So, it's a function of the shear stress at the boundary - can you get that from the model? Or I could calculate u_τ from the momentum flux at the boundary.

Indeed, if the shear stress at the wall is associated with a no-slip boundary condition, then it is calculated on-the-fly within the boundary tendency calculation, and is not stored anywhere. In order to use it for a subsequent computation, we'd have to re-compute it within a callback.

It also seems that one might want to, for example, average the momentum flux in horizontal / homogeneous directions before using it for this model constant calculation. But I'm not sure.

Either way, such a model is probably best implemented by pre-allocating an updatable object for the friction velocity (eg using `Ref`) and, using that updatable object within the `FunctionField` that computes the model constant, and then updating the object each time-step using a `Callback` within simulation.
